File: C++/1323.cpp
/*
    Author: Yuki
    GitHub: https://github.com/Yuki-14544869/
    Blog:   https://yuki-14544869.github.io/
*/
#include <bits/stdc++.h>
using namespace std;

const int  N = 1005;

int dp[N][N] = {0};

int min(const int a, const int b, const int c) {
    return min( a, min( b, c ) );
}

int main() {
    string input;
    cin >> input;
    int len = input.length();
    for (int i=len-1; i>=0; --i) {
        for (int j=i+1; j<len ; ++j) {
            if(input[i]==input[j])
                dp[i][j] = dp[i+1][j-1];
            else
                dp[i][j] = min( dp[i+1][j], dp[i][j-1], dp[i+1][j-1] ) + 1;
        }
    }
    cout << dp[0][len-1] << endl;
    return 0;
}

File: Java/1323.java
import java.util.Scanner;


public class Main {
    /*
    Author: Yuki
    GitHub: https://github.com/Yuki-14544869/
    Blog:   https://yuki-14544869.github.io/
    */
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        char[] s = in.next().toCharArray();
        int dp[][] = new int[1005][1005];
        int len = s.length;
        for(int i=len-1; i>=0; --i) {
            for(int j=i+1; j<len; ++j) {
                if(s[i] == s[j])
                    dp[i][j] = dp[i+1][j-1];
                else
                    dp[i][j] = Math.min(dp[i+1][j], Math.min(dp[i][j-1], dp[i+1][j-1]) ) + 1;
            }
        }
        System.out.println(dp[0][len-1]);
        in.close();
    }
}
